### ISSUE TITLE
Need to separate out org.apache.http shadows to compile against M developer preview

### ISSUE THREAD

**Issue by jhansche** [robolectric/robolectric#1862]
When setting `compileSdkVersion 'android-MNC'`, the `org.apache.http` classes are no longer on the compile classpath.  These classes have been deprecated since API level 22, and are now being removed completely in the M developer preview.

Attempting to compile a test class against Robolectric fails when calling `Robolectric.shadowOf(***)` (regardless of actual argument), because of:

```
error: cannot access AndroidHttpClient
```

There are other shadows that have been moved out into separate "plugin" dependency packages, so the same should be done for anything that references org.apache.http.

The other org.apache.http classes were moved out to the `shadows-httpclient` artifact;  but `android.net.http.AndroidHttpClient` was also removed, while the `ShadowAndroidHttpClient` class was not.

**Comment by erd** [comment]
`ShadowDefaultRequestDirector` already exists in the `shadows-httpclient` add-on and the `shadowOf` method should only exist in `org.robolectric.httpclient.Shadows`. Have you tested against 3.0-rc3?

**Comment by jhansche** [comment]
Wow, sorry.. I completely forgot that we had put our migration to 3.0 on hold (while the unit testing stuff in Gradle and AS was being fleshed out).  Resurrecting that migration did resolve the issue.

My bad :sob:

**Comment by jhansche** [comment]
Similar issue actually still exists even in 3.0-rc3.  This time it's [`android.net.http.AndroidHttpClient`](https://github.com/robolectric/robolectric/blob/master/robolectric-shadows/shadows-core/src/main/java/org/robolectric/shadows/ShadowAndroidHttpClient.java)

That class should probably move out to `shadows-httpclient`, because it is removed from android-MNC (along with all the other apache http classes).  This is only an issue when `compileSdkVersion = 'android-MNC'`

**Comment by erd** [comment]
We can move that class to a velocity template so it's only generated for API levels lower than M.

**Comment by hidroh** [comment]
Trying to target SDK 23 today when I encounter the same issue with `android.net.http.AndroidHttpClient`

**Comment by dawidgdanski** [comment]
Targetting API 23 causes same error as @hidroh mentions. Looks that SDK 23 totally got rid of `AndroidHttpClient`.

**Comment by jhansche** [comment]
@erd has there been any progress on the velocity template you mentioned?

**Comment by jbruchanov** [comment]
any progress with this issue ?

**Comment by erd** [comment]
We can't start work on this until Android M hits AOSP.

**Comment by jbruchanov** [comment]
workaround is to add it back...
gradle.build
```
android {
    compileSdkVersion 23
    buildToolsVersion "23.0.0"

    useLibrary 'org.apache.http.legacy'
    ....
}
```

**Comment by hidroh** [comment]
This is the official docs for @jbruchanov workaround: https://developer.android.com/preview/behavior-changes.html#behavior-apache-http-client

**Comment by jaredsburrows** [comment]
@jhansche  @hidroh Thanks for posting this!

**Comment by felipecsl** [comment]
... or you can add a class like this to your project. Works fine :)

```java
package android.net.http;

public class AndroidHttpClient {
}
```

**Comment by hidroh** [comment]
Thanks that's a very nice trick to make it compile!

**Comment by trevjonez** [comment]
The suggestion of @felipecsl is the best method. 
You can add this class in your projects `java/test` directory. By doing this, when your module gets built for production it isn't included in the APK/AAR.

If you use the android DSL option of adding `useLibrary 'org.apache.http.legacy'` you will be bloating your package with unused libraries. The android team removed the HttpClient for a reason, let's not circumvent that for only unit test config issues.

**Comment by jaredsburrows** [comment]
@felipecsl  @trevjonez Do your tests that rely on `AndroidHttpClient` pass when add the class in the `java/test` directory?

**Comment by felipecsl** [comment]
Nope, if you rely on AndroidHttpClient in any way, it will obviously not
work. This is just a hack to make the compiler happy

**Comment by jaredsburrows** [comment]
@felipecsl I was just talking about the tests. So I do not have add in  `useLibrary 'org.apache.http.legacy'`. Everything seems to work fine.

**Comment by gunhansancar** [comment]
@felipecsl thanks working.

**Comment by jaredsburrows** [comment]
@erd Can we close this?

**Comment by jhansche** [comment]
@erd @jaredsburrows Just want to point out that this shouldn't simply be "closed."  We should still fix it the way it was described [by @erd above](https://github.com/robolectric/robolectric/issues/1862#issuecomment-108681473), now that Marshmallow has hit AOSP.

**Comment by zsoltk** [comment]
+1 for fixing this. I'm not even using AndroidHttpClient in any way, and I can't run a robolectric test without either adding it with useLibrary (bad) or faking it out (bad).

**Comment by hameno** [comment]
+1 for moving it away from core

**Comment by erd** [comment]
Don't panic! Yes, this will need to be done to support API 23, which @jongerrish is working on.

**Comment by hk0i** [comment]
Is there a place I can see this change reflected to start using it or should I still hack it with an empty class?

**Comment by helinwang** [comment]
Another work around: use ` ((ShadowLooper) ShadowExtractor.extract(thread.getLooper()))` instead of `shadowof()`

**Comment by mattswitch** [comment]
@helinwang ShadowExtractor.extract() in place of Shadows.shadowOf() works.

**Comment by LiushuiXiaoxia** [comment]
Error:(26, 30) 错误: 无法访问AndroidHttpClient
找不到android.net.http.AndroidHttpClient的类文件

**Comment by LiushuiXiaoxia** [comment]
android {
    compileSdkVersion 23
    buildToolsVersion "23.0.0"

    useLibrary 'org.apache.http.legacy'
    ....
}

it's ok

**Comment by trevor-e** [comment]
Any updates on this?  M has been out for a while now.

**Comment by jongerrish** [comment]
It should be working for 3.1-SNAPSHOT

I think Mike Grafton was working on pushing out a 3.1 official release.

**Comment by trevor-e** [comment]
I was able to get it compiling by adding the blank `AndroidHttpClient` class, but now half my tests are failing with `Provider org.robolectric.shadows.gms.Shadows not a subtype` which looks like yet another open issue. :/  Tried both 3.0 and 3.1-SNAPSHOT.

**Comment by jongerrish** [comment]
What is the root cause? The subtype thing is a red herring, it occurs in
all subsequent tests when one test has failed during setup of the
environment. I'd check the first test failure.

Sorry in advance my ability to respond is pretty limited from where I am
now.

**Comment by trevor-e** [comment]
@jongerrish I was able to fix it, thanks for the help.  Knowing that the subtype error was a red herring clued me in to the real problem, which was Picasso crashing due to multiple singleton instances being set.

**Comment by itstexter** [comment]
Just pinging that this would be nice to have in 3.2 =]

**Comment by xian** [comment]
I believe this has been fixed for a while, please reopen if not.
